### ISSUE TITLE
Error: listen tcp x.x.x.x:443: bind: cannot assign requested address

### ISSUE THREAD

**Issue by noxware** [pocketbase/pocketbase#107]
When I try to run 

```
./pb serve --http="subdomain.domain.dev:80" --https="subdomain.domain.dev:443"
```

I get 

```
listen tcp x.x.x.x:443: bind: cannot assign requested address
```

I'm missing something?

**Comment by ganigeorgiev** [comment]
@noxware Is subdomain.domain.dev is a properly configured public domain? When using with `--https` we are issuing a Let's encrypt certificate and it requires the domain to be public/real (bare ips cannot be used too).

**Comment by noxware** [comment]
More info:

- The subdomain points to the device I want.
- TCP ports 80 and 443 are open.
- I can ping the subdomain and I get the correct IP.
- If I run `./pb serve --http="0.0.0.0:80"`, and write the IP to which the domain points (obtained by `ping`), I get the `index.html` I put in `pb_public`.
- I can't access to the domain directly in the browser without HTTPS because is a `.dev` domain. 
- Because of that, I configured certbot using [this steps](https://certbot.eff.org/instructions?ws=other&os=debianbuster).
- Certbot successfully configured the certificates giving the following output:

```console
$ sudo certbot certonly --standalone
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Please enter the domain name(s) you would like on your certificate (comma and/or
space separated) (Enter 'c' to cancel): X.X.dev
Requesting a certificate for X.X.dev

Successfully received certificate.
Certificate is saved at: /etc/letsencrypt/live/X.X.dev/fullchain.pem
Key is saved at:         /etc/letsencrypt/live/X.X.dev/privkey.pem
This certificate expires on 2022-10-12.
These files will be updated when the certificate renews.
Certbot has set up a scheduled task to automatically renew this certificate in the background.
```

- Then I tried running the command mentioned in the issue and got the following output:

```console
$ ./pb serve --http="X.X.dev:80" --https="X.X.dev:443"
> Server started at: https://X.X.dev:443
  - REST API: https://X.X.dev:443/api/
  - Admin UI: https://X.X.dev:443/_/
2022/07/15 02:43:56 listen tcp X.X.X.X:443: bind: cannot assign requested address
```

> Note: When you say `we are issuing a Let's encrypt certificate`, do you mean that pocketbase handles the process for you?

**Comment by ganigeorgiev** [comment]
> When you say we are issuing a Let's encrypt certificate, do you mean that pocketbase handles the process for you?
Yes, PocketBase will generate a Let's encrypt certificate for your domain when used with `--https` flag. There is no need for `certbot`.

You can still use `certbot` if you plan to put PocketBase behind nginx/apache, but then you don't need the `--https` flag.

**Comment by noxware** [comment]
> You can still use certbot if you plan to put PocketBase behind nginx/apache, but then you don't need the --https flag.

I was thinking about just using ngnix but now I'm curious about the problem.

> Does the subdomain has a valid A record pointing to your device/server?

Yeah, It's an `A + Dynamic DNS Record` (from NameCheap). It's pointing correctly to the device. 

> Can your device/server be accessed from outside of your local network?

Yeah, I can even ssh into the device by using the subdomain name.
While running `./pb serve --http="0.0.0.0:80"` I can do `wget http://X.X.dev/` and obtain the `index.html` that I put in `pb_public`. (I can't go trough the browser without `https` because is a `.dev` domain).

> Are you executing the serve command as root or with a user that is allowed to bind to the privileged ports?

It should be unnecessary since I used `setcap` as indicated in the docs. If that were the problem, I shouldn't be able to do  `./pb serve --http="0.0.0.0:80"` (which works for me).

I also tried with `sudo`, same result.

```console
$ sudo ./pb serve --http="X.X.dev:80" --https="X.X.dev:443"
> Server started at: https://X.X.dev:443
  - REST API: https://X.X.dev:443/api/
  - Admin UI: https://X.X.dev:443/_/
2022/07/16 00:05:56 listen tcp X.X.X.X:443: bind: cannot assign requested address
```

**Comment by noxware** [comment]
I also tried cross compiling a new version and running it with `sudo` but the same problem persists.

**Comment by ganigeorgiev** [comment]
I'm not sure how to debug it or reproduce it locally.

I would recommend try running `dig` just to ensure that everything is routed correctly.

If you server is a local device (eg. Raspberry Pi) maybe there could some missing/misconfigured port forwarding happening somewhere.

Maybe there is a firewall, iptables, etc, or maybe there is something interfering after the certbot installation, I'm not really sure.

Just for the tests, you could try installing nginx/apache and try issuing the certificate with certbot to see if that will work. If you are able to issue the certificate, then I really can't think of a reason why PocketBase can't do it.

**Comment by johnhorsema** [comment]
@noxware is the subdomain pointing to a non-local IP for your device? i.e. not `192.168.X.X`
also if your device has firewall or iptables, you need to set it to allow incoming TCP connections from port 443.

**Comment by noxware** [comment]
@johnhorsema  Yes, it's pointing to the device, using it's public IP.

If I run `./pb serve --http="0.0.0.0:80"` and then from my mobile phone, using the Mobile Network (not wifi) I try to wget the domain, everything works:.

![Screenshot_2022-07-16-13-58-21-500_com termux](https://user-images.githubusercontent.com/7684329/179365067-0e8171d7-3c2b-4e62-a463-e749d970ebeb.jpg)

In the screenshot you can see that I'm going trough a different network (not the local network), using the domain name (not the ip) and I'm getting the `index.html` file I mentioned before.

I can also try mounting a normal HTTP server (not HTTPS) using `./pb serve --http="0.0.0.0:443"` on port 443 and still works, which means that 443 is not blocked:

![image](https://user-images.githubusercontent.com/7684329/179365270-4aafaae8-4dae-4899-9eb1-c80e7a649b7b.png)

Extra info:
- The device is a Raspberry Pi 3 B+.
- The device has an static ip in the internal network.
- Ports in the router are configured to forward TCP connections on port 80 and 443 to that static IP.
- The public ip address is being updated using DDclient.
- Some other software in the device with access to network stuff are Docker and RaspAP (but it's webui is not running on port 80).
- The PocketBase binary has been cross compiled from a Windows machine to Linux arm32.

I don't think is a problem with ports. I can try configuring ngnix to confirm that it works if helps to discover the cause of this.
Additionally is there any log or any way to obtain detailed info about the cause of the error?

**Comment by ganigeorgiev** [comment]
@noxware Thanks for the detailed follow up.

Silly question - is the IP `186.52.x.x` from the screenshot your private ip or the public one managed by the ddclient? If ddclient uses the private one, then there is probably something wrong with its configuration.

To enable the dev errors logs in PocketBase, you could start the executable with `--debug`, but I'm not sure that this will be very helpful since the failure is during the application startup.

If nothing works, please test with nginx and certbot so that we can confirm that there is something wrong with PocketBase.

**Comment by noxware** [comment]
![image](https://user-images.githubusercontent.com/7684329/179368393-620b0a4d-afc3-49fd-87aa-bf5949438888.png)

It's definitely the public one.

I tried with `--debug` before but I remember that it only showed some sql stuff. I'm gonna try luck with ngnix in a moment.

**Comment by noxware** [comment]
Using ngnix as a reverse proxy with the location settings from the docs works.

```console
$ ./pb serve
> Server started at: http://localhost:8090
  - REST API: http://localhost:8090/api/
  - Admin UI: http://localhost:8090/_/
```

```
location / {
    # check http://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive
    proxy_set_header Connection '';
    proxy_http_version 1.1;
    proxy_read_timeout 180s;

    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto $scheme;

    proxy_pass http://localhost:8090;
}
```

![image](https://user-images.githubusercontent.com/7684329/179369465-6afd6561-2670-49ba-b770-e3dafaceab7d.png)

Weird.

**Comment by ganigeorgiev** [comment]
@noxware how did you run certbot? Just `sudo certbot --nginx -d yourdomain.dev`?

**Comment by noxware** [comment]
Originally I run `sudo certbot certonly --standalone` because I interpreted that PocketBase uses those files automagically.

Now, to configure the certificates for ngnix I just run `sudo certbot --nginx` which used the already existing certificates from the previous command and modified the ngnix configuration automatically. 

Seems strange to be an arm32 issue but who knows. If I find another sdcard and I get some free time I may try with a clean system installation and another subdomain.

For now I will stop investigating this. I was going to install ngnix anyway but the curiosity about this problem won me. 

Thanks for all the help!

**Comment by ganigeorgiev** [comment]
@noxware Actually this may not be related to the TLS generation.

If you find time and want to further investigate it, you could try starting the server with the domain name but with the `--http` flag only: `pocketbase serve --http="yourdomain.dev:80"`

I'm not familiar with ddclient (and setuping ddns in general), but if the above results in the same error, then my guess is that this something related to its configuration - maybe a missing loopback address, I don't know, and maybe it is worth checking your `/etc/hosts`. I'm not sure how nginx works in this case.

Anyway, thanks for spending the time debugging it. If anyone else is experiencing the same issue, please comment with information about your setup. I may try in the future playing with ddclient to emulate the setup.

**Comment by kennethklee** [comment]
I think I might be able to help clear this up.

- when you bind to `subdomain.domain.dev:443`, golang will resolve the DNS (only once) to get an IP and bind that IP. 
- `listen tcp x.x.x.x:443: bind: cannot assign requested address` - this error occurs when you try to bind to an IP that isn't local. in linux, you can show the list of local IPs with `hostname -I`.
- nginx will work because it binds to 0.0.0.0.
- certbot also works because it always binds to 0.0.0.0 in `--standalone`.

from the comments, `subdomain.domain.dev` doesn't resolve to a local IP. hence the error.

it's especially important to note that binding to a domain will not work with ddns, since golang will resolve the DNS only once.

here's some workarounds for @noxware:
1. keep nginx as gateway for SSL termination. manual certbot will work with nginx, ddclient will work, pocketbase will work. voila! alternatively, use caddy with Auto TLS -- `caddy --from subdomain.domain.dev --to localhost:8090` (top of my head, see their docs)
2. can setup loopback to public ip on the RPi. this is beyond pocketbase. simpler setup though.

I highly recommend caddy.

pocketbas can also tackle it by always binding to 0.0.0.0 for https. or maybe separate the https argument into listen and domain. who knows, up to them.

personally, I'd drop https support. luckily i'm not in the pocketbase team ;)

**Comment by ganigeorgiev** [comment]
I'll close this issue for now to cleanup a little the "backlog". As @kennethklee mentioned binding to `0.0.0.0` or adding a loopback address may have resolved @noxware issue. If anyone is experiencing similar issue, please comment and I'll reopen it.

For now, I'll not change the PocketBase default to `0.0.0.0` since it may not be desirable to listen on every available network interface, but I'll consider for future versions to optionally split the tls issuing domain from the listening address so that users can specify it explicitly.

**Comment by doniyorz** [comment]
HI all. I'm trying to start pocketbase on VPS using 
`./pocketbase serve --http="subdomain.domain.dev:80" --https="subdomain.domain.dev:443"`
and it says
 `listen tcp `{server's public ip}`:80: bind: cannot assign requested address`
 to both 80 and 443 ports. Domain is public and properly configured, pointing out to my VPS.  I haven't tried using caddy or setting up loopback to public IP, just wanted to mention that it happens not only on Raspberry. I'll write a follow up comment after trying caddy or loopback setup

**Comment by ganigeorgiev** [comment]
@doniyorz Make sure that you can bind to the privileged 80 and 443 ports (eg. as root) and that you have a proper loopback address configured in your `/etc/hosts`.

**Comment by doniyorz** [comment]
It binds both 80 and 443 with no problem when I try to serve on 0.0.0.0, and I can connect by my VPS address from outside, but https doesn't work obviously.

**Comment by ganigeorgiev** [comment]
@doniyorz Then you are most likely missing a loopback address. If you don't want to setup one, you can always put PocketBase behind a reverse proxy like nginx/apache/caddy/etc. You could find an example nginx configuration in https://pocketbase.io/docs/going-to-production/#deploy-behind-reverse-proxy.

**Comment by doniyorz** [comment]
I've added loopback address, now it starts ok but I cant reach it from outside neither by IP or domain

**Comment by ganigeorgiev** [comment]
@doniyorz Are you sure that you've set the correct address? Could you provide some inform on your setup, like what VPS provider do you use, what is the content of your `/etc/hosts`, how did you setup the domain, are you using IPv4 or IPv6, etc.?

**Comment by doniyorz** [comment]
I'm sorry, I'm kinda new to setting up linux servers. I figured it out. I've checked ifconfig and got two entries there,  one of which had ` UP,BROADCAST,RUNNING,MULTICAST` flags and second  `UP,LOOPBACK,RUNNING`. So I got the ip address from the first one and added it to `/etc/hosts` like this `ip-address domain`. Then I ran `./pocketbase serve --http="subdomain.domain:80" --https="subdomain.domain:443"` and it worked. I checked it in browser and everything was good. Sorry for wasting your time and thank you for the great work!

**Comment by whkoh** [comment]
I have a caddy config working serving multiple other applications. 

trying to add pocketbase, but the configs I'm trying aren't working.

I've tried `./pocketbase serve --https="0.0.0.0:8090"` suggested above, as well as:
- `./pocketbase serve --https="0.0.0.0:8090" --http="0.0.0.0:8090"`
- `./pocketbase serve --https="127.0.0.1:8090" --http="127.0.0.1:8090"`
- `./pocketbase serve --https="192.168.66.100:8090" --http="192.168.66.100:8090"`
    - http works but https does not. error: `2022/10/09 16:17:53 http: TLS handshake error from 192.168.50.244:51096: acme/autocert: missing server name`

caddy is doing automatic TLS with letsencrypt, the config is as follows. I don't see any unusual error in caddy logs.

```
pb.my.domain {
        route {
                import eppsecure
                reverse_proxy 192.168.66.100:8090
        }
}
```

**Comment by ganigeorgiev** [comment]
@whkoh If caddy is handling the TLS then you don't need to set the `--https` flag (additionally Let's encrypt doesn't support issuing domains on bare IPs so providing an IP as value to the https flag will not work).

**Comment by raf2001** [comment]
now I'm confused...I have clean installation (on arm infrastructure) with ubuntu 22.04 and still after adding ip to hosts I get :
2022/10/12 22:11:24 listen tcp 130.61.x.x:443: bind: cannot assign requested address
```
my /etc/hosts:
127.0.0.1       localhost
# The following lines are desirable for IPv6 capable hosts
::1     ip6-localhost   ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
ff02::3 ip6-allhosts
127.0.1.1       db      db
130.61.x.x  db.x.pl      db.x.pl
130.61.x.x  db      db
```

and ufw:
```
root@db:/home/ubuntu# ufw status
Status: active

To                         Action      From
--                         ------      ----
80                         ALLOW       Anywhere
443                        ALLOW       Anywhere
8090                       REJECT      Anywhere
80 (v6)                    ALLOW       Anywhere (v6)
443 (v6)                   ALLOW       Anywhere (v6)
8090 (v6)                  REJECT      Anywhere (v6)
```


There isn't any other app (like nginx) on this machine. 
On dns I have configured A record with content (yes it's cloudflare)
A | db | 130.61.x.x | DNS only | Auto

and also tried to start PB by:
./pocketbase serve --http="db.x.pl:80" --https="db.x.pl:443"
and
./pocketbase serve --http="130.61.x.x:80" --https="130.61.x.x:443"
with no result.

**Comment by bacon-delight** [comment]
I'm also having the same issue, both 80 and 443 ports can't be assigned. I'm using DigitalOcean Droplet and I've done the following:
- Using NGINX Config as:
```
server {
    listen 80;
    server_name <url>;

    location / {
        proxy_set_header Connection '';
        proxy_http_version 1.1;
        proxy_read_timeout 360s;

        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_pass http://127.0.0.1:8090;
    }
}
```

- DNS Subdomain with A Name pointing to IP (tested with the basic NGINX page before the above config was added)
- `./pocketbase serve` works and I can access the admin panel

But when I do `/home/<user>/pocketbase serve --http="dev.contrimoney.com:80" --https="dev.contrimoney.com:443"` or `/home/<user>/pocketbase serve --http="dev.contrimoney.com:80"`, I get `bind: cannot assign requested address` error for both port 80 and 443 from both root and non-root user

### PLEASE HELP!

**Comment by ganigeorgiev** [comment]
@bacon-delight Why do you use nginx in the first place? If you have nginx running and already listening on 80 and 443 ports, you can't bind another process to the same ports (_unless you use `SO_REUSEPORT` or similar_).

If you put PocketBase behind a reverse proxy (nginx), you have to start the process with the address and port you've in your configuration. In your case that is `127.0.0.1:8090`. Or in other words:

```
/home/<user>/pocketbase serve --http="127.0.0.1:8090"
```

If you use nginx, then you must use certbot since the PocketBase auto Let's encrypt issuing will not work (or at least not without hacky workarounds).

**Comment by raf2001** [comment]
ok. now it's working. There was couple issue with OracleCloud (adding http/https port on traffic in network configuration tab). 
Coupke hours later finally I can start to do something with that.

**Comment by jimafisk** [comment]
I'm hitting the `bind: cannot assign requested address` error as well using a Digital Ocean droplet. If I `vim /etc/hosts` and add an entry like `127.0.0.1 sub.mydomain.com` it _does_ lets me bind to the address with `/root/pb/pocketbase serve --http="sub.mydomain.com:80" --https="sub.mydomain.com:443"` but if I try to visit the admin in the browser it says "This site can’t be reached."

I assume I'm not doing the loopback address correctly? I've tried with several internal IPs but the same thing happens and if I try with the public IP that I'm pointing my A Record at, I can't bind at all (the original issue).

I have a `ufw status` of:
```
80/tcp                     ALLOW       Anywhere                  
443/tcp                    ALLOW       Anywhere                  
80/tcp (v6)                ALLOW       Anywhere (v6)             
443/tcp (v6)               ALLOW       Anywhere (v6)
```

And I can `ping sub.mydomain.com` successfully.

@doniyorz do you have any other hints on how you ultimately resolved this? I tried all the IPs returned from `ifconfig`. Thanks!

**Comment by ganigeorgiev** [comment]
@jimafisk

The reason for the above error most likely is because the PocketBase server tries to bind only to the IP that is resolved from the domain (aka. we don't bind to `0.0.0.0` or `[::]`). And this is why for public facing machines the loopback address usually should be with the IP of the host (either from `ifconfig`, `hostname -I`, etc.).

But with DigitalOcean (and probably other VPS providers) manually configuring this usually shouldn't be necessary since the `/etc/hosts` I believe is "managed" by default and the entries will be added automatically when setting the domain from their interface (usually there should be a comment at the start of the hosts file that points to `vps.cfg`, `cloud.cfg` or similar, and I'm not sure if in this case manually editing the `/etc/hosts` will persist).

**Comment by jimafisk** [comment]
Thank you @ganigeorgiev that worked! :heavy_check_mark: 

I ran `hostname -I` and that returned several IP addresses. The first was the ip4 address that I could see from the Digital Ocean dashboard that I had already tried unsuccessfully, but the second was an internal IP `10.10.0.6` that worked! Just needed to `vim /etc/hosts`:

```
10.10.0.6 sub.mydomain.com
```

Just a couple of notes:
- I'm not managing DNS through Digital Ocean, I just created an A Record from a different registrar and pointed it to the "Reserved IP" I set up for my droplet
- I restarted my droplet from the Digital Ocean dashboard and my custom `/etc/hosts` entry is still intact
- Restarting the server allowed the `systemctl` settings [you described](https://github.com/pocketbase/pocketbase/discussions/512#discussioncomment-3668130) to work. Before the restart I could `/root/pb/pocketbase serve --http="sub.mydomain.com:80" --https="sub.mydomain.com:443"` manually to start PocketBase but I couldn't `systemctl enable pocketbase.service --now` to get it working in a persistent manner.

**Comment by harryc9** [comment]
Had the same issue of `bind: cannot assign requested address` when trying to serve with `--https`

Resolved with @ganigeorgiev and @jimafisk suggestions for adding network address IP to `/etc/hosts`.
No caddy/nginx used 👍

**Comment by adampatterson** [comment]
> ok. now it's working. There was couple issue with OracleCloud (adding http/https port on traffic in network configuration tab). Coupke hours later finally I can start to do something with that.

I'm also on the Oracle cloud, I was able to bind without any issues on my Digital Ocean droplet but the Oracle VPS very strange.

what I am slowing figuring out as I stumble through areas of lunux that I have no bunsess being in is that the public IP is not attached to the VPS.

If I were to run `ifconfig -a | grep "inet"` on a Digital Ocean VPS you will see the public IP.

Running it on the Oracle Cloud Compute doesn't show it at all. You will need to bind to the Private IP address.

**Comment by mingyeungs** [comment]
I encountered the same issue on AWS EC2 (Ubuntu 22), and @jimafisk ’s reply solved my issue.

At first:

```
sudo /root/pb/pocketbase serve --http="example.com:80"
# > Server started at: http://example.com:80
# > ...
# > listen tcp 43.xx.xx.xx:80: bind: cannot assign requested address
```

The fix:

```
hostname -I
# this returns the EC2’s private IP address e.g. 172.xx.xx.xx

sudo vi /etc/hosts
# add `172.xx.xx.xx  example.com`
```

Afterward:

```
sudo /root/pb/pocketbase serve --http="example.com:80"
# success!
```

To avoid using sudo, use `setcap` as described in [official guide](https://pocketbase.io/docs/going-to-production/#minimal-setup)

```
sudo setcap 'cap_net_bind_service=+ep' /root/pb/pocketbase

/root/pb/pocketbase serve --http="example.com:80"
# success!
```

Adding `--https=...` generate a number of warning/error on the first run,
though HTTPS does work in the browser.

```
/root/pb/pocketbase serve --http="example.com:80" --https="example.com:443"
# 2023/01/20 05:42:51 http: TLS handshake error from 139.xx.xx.xx:55094: acme/autocert: host "43.xx.xx.xx" not configured in HostWhitelist
# ...
# 2023/01/20 05:42:52 http: TLS handshake error from 139.xx.xx.xx:56340: tls: client requested unsupported application protocols ([http/0.9 http/1.0 spdy/1 spdy/2 spdy/3 h2c hq])
# 2023/01/20 05:42:52 http: TLS handshake error from 139.xx.xx.xx:56794: tls: client requested unsupported application protocols ([hq h2c spdy/3 spdy/2 spdy/1 http/1.0 http/0.9])
# 2023/01/20 05:42:53 http: TLS handshake error from 139.xx.xx.xx:57104: acme/autocert: host "43.xx.xx.xx" not configured in HostWhitelist
# ...
# 2023/01/20 05:52:24 http: TLS handshake error from 43.xx.xx.xx:41340: acme/autocert: missing server name
```


According to AWS [doc](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-instance-addressing.html#concepts-private-addresses), the private IP would not change unless the instance is terminated so this fix should keep on working(?):

”A private IPv4 address, regardless of whether it is a primary or secondary address, remains associated with the network interface when the instance is stopped and started, or hibernated and started, and is released when the instance is terminated.”

**Comment by AriaFantom** [comment]
> ok. now it's working. There was couple issue with OracleCloud (adding http/https port on traffic in network configuration tab). Coupke hours later finally I can start to do something with that.

Can you please explain the steps that you followed to make it work

**Comment by adampatterson** [comment]
@AriaFantom If I remember correctly I used the private IP `./pocketbase serve --http="PRIVATEIP:8080"

You will need to head over to your Default security list for the Oracle Cloud VNC and open up port 8080 in this case.

1. Networking
2. Virtual cloud network
3. Click the link in the table  ( vnc-numbers-numbers )
4. Click the link in the table ( Default security group ) 
5. Add an ingress rule for 8080

I don't know if you also need to run 
```shell
sudo iptables -I INPUT 6 -m state --state NEW -p tcp --dport 8080 -j ACCEPT
sudo netfilter-persistent save
```

 I think I Googled this part and it seemed to work 🤷‍♂️

Now you should be able to access the site using `http://PRIVATEIP:8080`



But I settled on using Nginx with Proxy Pass so that I could access it through a domain and HTTPS.

Use cloud flare, create a subdomain like `db.example.com`

`./pocketbase serve --http="127.0.0.1:8080"`

```
server {
    listen 80;
    server_name db.example.com;
    client_max_body_size 10M;

    location / {
        # check http://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive
        proxy_set_header Connection '';
        proxy_http_version 1.1;
        proxy_read_timeout 360s;

        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_pass http://127.0.0.1:8080;
    }
}
```

How you should be able to access your DB through https://db.example.com

I'd also recommend reading this https://pocketbase.io/docs/going-to-production/


Hope that helps

**Comment by Hammaad-M** [comment]
I've tried everything shared above to setup pocketbase on an Azure VM, and still get `bind : address already in use`. First, I tried just running the executable and received the error above. I then setup a reverse proxy with Nginx to listen at port 80 and forward all requests to 127.0.0.1:8090 where I ran the executable--but can't access the admin dashboard at `<public_ip>:80/_`.

I got the private IP using `hostname -I`, which is `10.0.0.4`. I added the following line to etc/hosts: `10.0.0.4 <public_ip>`. Tried to run executable at <public_ip>:80, swapping 80 with every port I know, but get `bind: address already in use`. Then tried running at `10.0.0.4:80, which ran successfully but admin dashboard is not accessible. 

Essentially, I can't run the executable on the public IP, but can run it on the local IP--only then I can't access the admin dashboard from the browser. Using Nginx to proxy and the hosts file is simply not working. I'm all out of ideas. I even tried starting over on a google VM instance, but get the same error. 

Any help would be much appreciated...

**Comment by ganigeorgiev** [comment]
@Hammaad-M Please don't spam old discussion and instead try to open a new one with more details about your current deployment setup, including:

- the exact steps and commands you've performed
- the configuration file of your reverse proxy (if you use one)

In any case, if you get `bind : address already in use` then this most likely means that there is another application already running on the same address:port.

If you have more than one application on the server and want all of them to bind to 80/443 then you'll need a reverse proxy (nginx, apache, caddy, etc.). For example nginx config, please check https://pocketbase.io/docs/going-to-production/#using-reverse-proxy.

Also if you have a public domain name and want to bind to 80/443 then you don't have to specify any IP address. The above instructions are a little outdated (see https://github.com/pocketbase/pocketbase/discussions/3190). Or in other words, it should be enough to run:
```
./pocketbase serve yourdomain.com
```

**Comment by Hammaad-M** [comment]
@ganigeorgiev, apologies for cluttering up old threads, I've posted a new issue [here](https://github.com/pocketbase/pocketbase/issues/3937).
